routing/payment_state_machine.go

```diff
@@ -0,0 +1,467 @@
+package routing
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/coreos/bbolt"
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/htlcswitch"
+	"github.com/lightningnetwork/lnd/lntypes"
+)
+
+var (
+	// paymentsBucketKey is the key used to access the top level bucket of
+	// the payment state machine. This bucket contains sub-buckets indexed
+	// by the payment hash of the active payments.
+	paymentsBucketKey = []byte("router-payment-state-machine")
+
+	// paymentStateKey is a key used in an active payment bucket to access
+	// the current payment state.
+	paymentStateKey = []byte("payment-state-key")
+
+	// paymentAttemptKey is used to access the information of the current
+	// payment attempt sent to the switch. This key is only populated when
+	// the payment is in the state paymentStateAttemptSent.
+	paymentAttemptKey = []byte("payment-attempt-key")
+
+	// paymentErrorKey is a key that accesses the last encountered error
+	// for payment attempts for the active payment. It is populated when
+	// the active payment is in the state paymentStateFailed, and sometimes
+	// in other states if the previous payment attempt failed.
+	paymentErrorKey = []byte("payment-error-key")
+
+	// ErrNotFound is an error returned if the given payment hash was not
+	// found.
+	ErrNotFound = errors.New("payment hash not found")
+)
+
+// paymentStateMachine is the persistent state machine used by the
+// ChannelRouter to track the progress of active payments. It is used to keep
+// information across restarts, needed to properly retry failed paymets, and
+// handle payment errors and successes encountered for a payment sent before
+// the ChannelRouter was restarted.
+type paymentStateMachine struct {
+	db *channeldb.DB
+}
+
+// paymentState is the type used to indicate which state a given payment has in
+// the payment state machine.
+type paymentState byte
+
+const (
+	// paymentStateIntended is the starting state of a payment. A payment
+	// in this state is intended for sending, but haven't yet had a
+	// paymentAttempt sent to the Switch..
+	paymentStateIntended paymentState = 1
```

Comment:
Is this state really needed? What would go wrong if we find out that a payment is already in flight when we try to set the state to `paymentStateAttemptSent` ?

Reply:
The sequence `paymentStateAttemptSent` -> `paymentStateIntended` -> `paymentStateAttemptSent` after a failure, does it need to be persisted? Or can we just overwrite the route?

Reply:
The intermediate error can possibly just stay in memory. And hopefully we will remove it alltogether soon, because it is of very limited use in the context of a payment.

Reply:
> Is this state really needed? What would go wrong if we find out that a payment is already in flight when we try to set the state to `paymentStateAttemptSent` ?

We need this state since it is only here where we safely can cancel, timeout, fail the payment. If we only store `paymentStateAttemptSent`, we would need to ask the `switch` about the result, then do the cancelation logic before trying again. I found it easier to reason about when  having an explicit state where this is safe.

> The intermediate error can possibly just stay in memory. And hopefully we will remove it alltogether soon, because it is of very limited use in the context of a payment.

It has the nice benefit of keeping the current behavior of returning the last encountered error. Maybe we want to remove it later, but I think that's a different discussion.